From 978262a9980e2939448bad9f2c7434605f01f778 Mon Sep 17 00:00:00 2001
From: Wenzong Fan <wenzong.fan@windriver.com>
Date: Mon, 7 Mar 2016 22:24:32 -0500
Subject: [PATCH] wr-policy: fix su failure with non-root users

Reference sources: selinux-policy-3.13.1-128.21.fc22.src.rpm
Changes come from: policy-f22-base.patch

* Allow *_su_t search /run/systemd/journal (mls_systemhigh)
* Allow *_su_t create connections to the system bus
* Allow domain the attributes of selinuxfs filesystem
* Allow systemd_logind_t read the process state (/proc/pid)
  of all domains

Fix avc denials:

  avc: denied { search } for pid=888 comm="su" \
  name="journal" dev="tmpfs" ino=8203 \
  scontext=staff_u:sysadm_r:sysadm_su_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:syslogd_var_run_t:s15:c0.c1023 \
  tclass=dir permissive=1

  avc: denied { search } for pid=888 comm="su" \
  name="dbus" dev="tmpfs" ino=13344 \
  scontext=staff_u:sysadm_r:sysadm_su_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:system_dbusd_var_run_t:s0 \
  tclass=dir permissive=1

  avc: denied { write } for pid=888 comm="su" \
  name="system_bus_socket" dev="tmpfs" ino=13345 \
  scontext=staff_u:sysadm_r:sysadm_su_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:system_dbusd_var_run_t:s0 \
  tclass=sock_file permissive=1

  avc: denied { connectto } for pid=888 comm="su" \
  path="/run/dbus/system_bus_socket" \
  scontext=staff_u:sysadm_r:sysadm_su_t:s0-s15:c0.c1023 \
  tcontext=system_u:system_r:system_dbusd_t:s0-s15:c0.c1023 \
  tclass=unix_stream_socket permissive=1

  avc: denied { getattr } for pid=893 comm="su" \
  name="/" dev="selinuxfs" ino=1 \
  scontext=staff_u:sysadm_r:sysadm_su_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:security_t:s0 \
  tclass=filesystem permissive=0

  avc: denied { search } for pid=337 comm="systemd-logind" \
  name="880" dev="proc" ino=17949 \
  scontext=system_u:system_r:systemd_logind_t:s0-s15:c0.c1023 \
  tcontext=staff_u:sysadm_r:sysadm_su_t:s0-s15:c0.c1023 \
  tclass=dir permissive=1

  avc: denied { read open getattr } for pid=337 comm="systemd-logind" \
  name="cgroup" dev="proc" ino=17981 \
  scontext=system_u:system_r:systemd_logind_t:s0-s15:c0.c1023 \
  tcontext=staff_u:sysadm_r:sysadm_su_t:s0-s15:c0.c1023 \
  tclass=file permissive=1

Upstream-Status: Pending

Signed-off-by: Wenzong Fan <wenzong.fan@windriver.com>
---
 policy/modules/admin/su.if       | 4 ++++
 policy/modules/kernel/domain.te  | 3 +++
 policy/modules/system/systemd.te | 2 ++
 3 files changed, 9 insertions(+)

--- refpolicy.orig/policy/modules/admin/su.if
+++ refpolicy/policy/modules/admin/su.if
@@ -206,6 +206,10 @@ template(`su_role_template',`
 	init_rw_utmp($1_su_t)
 
 	mls_file_write_all_levels($1_su_t)
+	mls_file_read_all_levels($1_su_t)
+
+	# creating connections to the system bus
+	dbus_system_bus_client($1_su_t)
 
 	logging_send_syslog_msg($1_su_t)
 
--- refpolicy.orig/policy/modules/kernel/domain.te
+++ refpolicy/policy/modules/kernel/domain.te
@@ -122,6 +122,9 @@ files_list_root(domain)
 # Yocto/oe-core use some var volatile links
 files_read_var_symlinks(domain)
 
+# get the attributes of the selinuxfs filesystem
+selinux_getattr_fs(domain)
+
 ifdef(`hide_broken_symptoms',`
 	# This check is in the general socket
 	# listen code, before protocol-specific
--- refpolicy.orig/policy/modules/system/systemd.te
+++ refpolicy/policy/modules/system/systemd.te
@@ -323,6 +323,8 @@ fs_mount_tmpfs(systemd_logind_t)
 auth_manage_var_auth(systemd_logind_t)
 mls_file_read_all_levels(systemd_logind_t)
 
+domain_read_all_domains_state(systemd_logind_t)
+
 userdom_manage_all_user_tmp_content(systemd_logind_t)
 userdom_mounton_tmp_dirs(systemd_logind_t)
 
